src/pykeen/models/predict.py

```diff
@@ -115,6 +114,143 @@ def _get_input_batch(
+
+        :param df: nrows: n
+            the predictions
+        :param mapped_triples: shape: (m, 3)
+            the ID-based triples
+        :param invert:
+            whether to invert the result
+
+        :return: shape: (n,), dtype: bool
+            a boolean mask indicating which row is contained in the given ID-based triples
+        """
+        raise NotImplementedError
+
+    def filter(self, df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Filter out known triples.
+
+        :param df:
+            the predictions
+
+        :return:
+            the filtered dataframe
+        """
+        for mapped_triples in self.filter_triples.values():
+            df = df[self._contains(df=df, mapped_triples=mapped_triples, invert=True)]
+        return df
+
+    def contains(self, df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Add column indicating whether the triples are known.
+
+        :param df:
+            the predictions
+
+        :return:
+            the predictions with extra columns
+        """
+        for key, mapped_triples in self.filter_triples.items():
+            df[f"in_{key}"] = self._contains(df=df, mapped_triples=mapped_triples)
+        return df
+
+    def process(self, df: pd.DataFrame, remove_known: bool, add_novelties: bool) -> pd.DataFrame:
+        """Post-process a prediction dataframe."""
+        if add_novelties and not remove_known:
+            return self.contains(df=df)
+        elif remove_known:
+            return self.filter(df=df)
+        return df
+
+
+class SinglePredictionPostProcessor(PredictionPostProcessor):
+    """Post-processor for single-target predictions."""
+
+    def __init__(self, target: Target, other_columns_fixed_ids: Tuple[int, int], **kwargs) -> None:
+        """Initialize the post-processor.
+
+        :param target:
+            the prediction target
+        :param other_columns_fixed_ids:
+            the fixed IDs for the other columns
+
+        :param kwargs:
+            additional keyword-based parameters passed to :meth:`PredictionPostProcessor.__init__`
+        """
+        super().__init__(**kwargs)
+        self.target = target
+        self.other_columns_fixed_ids = other_columns_fixed_ids
+
+    # docstr-coverage: inherited
+    def _contains(
+        self, df: pd.DataFrame, mapped_triples: MappedTriples, invert: bool = False
+    ) -> numpy.ndarray:  # noqa: D102
+        col = TARGET_TO_INDEX[self.target]
+        other_cols = sorted(set(range(mapped_triples.shape[1])).difference({col}))
+        device = mapped_triples.device
+        other_col_ids = torch.as_tensor(data=self.other_columns_fixed_ids, dtype=torch.long, device=device)
+        filter_mask = (mapped_triples[:, other_cols] == other_col_ids[None, :]).all(dim=-1)
+        known_ids = mapped_triples[filter_mask, col].unique()
+        query_ids = torch.as_tensor(df[f"{self.target}_id"].to_numpy(), device=mapped_triples.device)
```

Comment:
is there a reason for `device=mapped_triples.device` instead of `device=device`?

Reply:
https://github.com/pykeen/pykeen/commit/c3ac29fc7eb9ecce8a620ccb41b499fdf5f35919